File: knowmind-web/src/services/workspaceFiles.ts
/**
 * 受控文件工作区 API（需登录）。
 */

import { apiFetch } from "@/services/http";

const PREFIX = "/workspace/files";

export type AllowedRootsResponse = {
  allowed_roots: string[];
  hint_env: string;
};

export type FileReadResponse = {
  path: string | null;
  content: string | null;
  status: string;
  truncated?: boolean;
  size_bytes?: number;
};

export type FileWriteResponse = {
  path: string | null;
  status: string;
  bytes_written?: number;
};

export async function fetchAllowedRoots(): Promise<AllowedRootsResponse> {
  const res = await apiFetch(`${PREFIX}/roots`);
  if (!res.ok) throw new Error(await res.text());
  return (await res.json()) as AllowedRootsResponse;
}

export async function readWorkspaceFile(path: string): Promise<FileReadResponse> {
  const res = await apiFetch(`${PREFIX}/read`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ path }),
  });
  if (!res.ok) throw new Error(await res.text());
  return (await res.json()) as FileReadResponse;
}

export async function writeWorkspaceFile(
  path: string,
  content: string,
  options?: { format?: "auto" | "markdown" | "text"; overwrite?: boolean },
): Promise<FileWriteResponse> {
  const res = await apiFetch(`${PREFIX}/write`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      path,
      content,
      format: options?.format ?? "auto",
      overwrite: options?.overwrite ?? true,
    }),
  });
  if (!res.ok) throw new Error(await res.text());
  return (await res.json()) as FileWriteResponse;
}

File: knowmind-server/app/api/v1/endpoints/workspace_files.py
from fastapi import APIRouter, Depends, HTTPException

from app.api.deps import get_current_user_id
from app.core.config import settings
from app.models.workspace_schemas import (
    AllowedRootsResponse,
    FileOpResponse,
    FileReadRequest,
    FileWriteRequest,
)
from app.services import file_workspace

router = APIRouter()


def _ensure_file_workspace() -> None:
    if not settings.file_tools_enabled:
        raise HTTPException(status_code=403, detail="服务端未启用文件工作区")


@router.get("/roots", response_model=AllowedRootsResponse)
async def list_roots(_user_id: str = Depends(get_current_user_id)) -> AllowedRootsResponse:
    _ensure_file_workspace()
    data = file_workspace.list_allowed_roots_payload()
    return AllowedRootsResponse(**data)


@router.post("/read", response_model=FileOpResponse)
async def read_file(
    body: FileReadRequest,
    _user_id: str = Depends(get_current_user_id),
) -> FileOpResponse:
    _ensure_file_workspace()
    try:
        data = file_workspace.read_document(body.path, max_bytes=file_workspace.max_read_bytes())
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return FileOpResponse(
        path=data.get("path"),
        content=data.get("content"),
        status=str(data.get("status", "read")),
        truncated=bool(data.get("truncated")),
        size_bytes=data.get("size_bytes"),
    )


@router.post("/write", response_model=FileOpResponse)
async def write_file(
    body: FileWriteRequest,
    _user_id: str = Depends(get_current_user_id),
) -> FileOpResponse:
    _ensure_file_workspace()
    try:
        data = file_workspace.write_document(
            body.path,
            body.content,
            format=body.format,  # type: ignore[arg-type]
            overwrite=body.overwrite,
        )
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e
    return FileOpResponse(
        path=data.get("path"),
        status=str(data.get("status", "written")),
        bytes_written=data.get("bytes_written"),
    )
